### ISSUE TITLE
named constants in expressions?

### ISSUE THREAD

**Issue by JaapAap** [kaitai-io/kaitai_struct#174]
I am interpreting a binary format that contains integers that have to be scaled by factors given by (mostly negative) powers of two, so 2^-43 (1.136868377216160e-13), 2^-31 (4.656612873077393e-10) etc. The feature of using scaled instances is really nice to allow this scaling, however, the constants are hard to read. It would be nice to be able to used named constants instead. Is this possible at all?

**Comment by JaapAap** [comment]
Also, it seems like the compiler does not like constants in exponential format in the first place.... Although 1e-5 is excepted, 1.1e-5 is not. Would it be possible to fix that?

**Comment by GreyCat** [comment]
Named constants can be generally defined as value instances, i.e.:

```yaml
instances:
  foo:
    value: 42
```

In fact, all instances are supposed to be calculated / parsed only once, so they are already "constants" in that aspect.

Thanks for noting that we have problem with exponential floats parsing, I'll try to look into it. In fact, this is hard to find, as doing stuff like that:

```yaml
instances:
  not_quite_pi:
    value: 3.1415926e-2
```

actually works, as in this case the number is parsed by YAML parser, not by our language parser.

**Comment by JaapAap** [comment]
Thanks - that indeed works but I don't think that (part of) the instance section can be imported from another location to allow re-use of these constants?

**Comment by GreyCat** [comment]
Import per se will work, but in general, we have very little support for addressing individual subtypes in anything besides `type` and typecasting (i.e. in enums, accessing any non-instance class members (AKA "static"), singletons, etc). So, adding extra "constants" definition area won't make it accessible automatically anyway.

Please tell a bit more about what you have in mind?

**Comment by JaapAap** [comment]
It would be a nice-to-have facility to have a file with constants
```
Meta:
    Id: powers_of_two
Instances:
  two_to_the_power_2:
    value: 4
  two_to_the_power_min_2:
    value: 0.25
```
and then importing these by listing the file in the  `imports: ` section of my data definitions  to allow writing instances such as unscaled_number*two_to_the_power_min_2.

**Comment by JaapAap** [comment]
If you could point me to the place where this is parsed, I'd like to give it a try too (even though my Scala skills are very limited)

**Comment by GreyCat** [comment]
If that's all you want, actually you can do it right now with something like that:

```yaml
# powers_of_two.ksy
meta:
  id: powers_of_two
instances:
  two_to_the_power_2:
    value: 4
  two_to_the_power_min_2:
    value: 0.25

# your_actual_file.ksy
meta:
  id: your_actual_file
  imports:
    - powers_of_two
seq:
  - id: some_value
    type: s1
instances:
  consts:
    type: powers_of_two
  mod_value:
    value: some_value * consts.two_to_the_power_min_2
```

This comes at expense of one extra object (like singleton) created in memory for all these constants.

**Comment by JaapAap** [comment]
Thanks - that is quite a nice solution actually!

**Comment by GreyCat** [comment]
I guess I should it this trick into a user guide then ;)
And I'll start a separate issue about parsing of exponentian literals.
